src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.cs

```diff
@@ -155,27 +146,23 @@ public static JsonElement CreateJsonSchema(
         serializerOptions ??= DefaultOptions;
         inferenceOptions ??= AIJsonSchemaCreateOptions.Default;
 
-        FunctionParameterKey key = (
+        SchemaGenerationKey key = new(
             type,
-            ParameterName: null,
+            parameterName: null,
             description,
             hasDefaultValue,
             defaultValue,
-            inferenceOptions.IncludeSchemaKeyword,
-            inferenceOptions.DisallowAdditionalProperties,
-            inferenceOptions.IncludeTypeInEnumSchemas,
-            inferenceOptions.RequireAllProperties,
-            inferenceOptions.FilterDisallowedKeywords);
+            inferenceOptions);
 
         return GetJsonSchemaCached(serializerOptions, key);
     }
 
-    private static JsonElement GetJsonSchemaCached(JsonSerializerOptions options, FunctionParameterKey key)
+    private static JsonElement GetJsonSchemaCached(JsonSerializerOptions options, SchemaGenerationKey key)
     {
         options.MakeReadOnly();
-        ConcurrentDictionary<FunctionParameterKey, JsonElement> cache = _schemaCaches.GetOrCreateValue(options);
+        ConcurrentDictionary<SchemaGenerationKey, JsonElement> cache = _schemaCaches.GetOrCreateValue(options);
 
-        if (cache.Count >= CacheSoftLimit)
+        if (cache.Count >= CacheSoftLimit || key.TransformSchemaNode is not null)
```

Comment:
cache.Count takes all the locks in the CD and is not cheap. It'd be a bit better to do the TransformSchemaNode check first.

Reply:
True, although that would only be optimizing for the case where a transformer is specified, which does not cache and is therefore slow by definition.

Reply:
Yes, but in the common case, both of these conditions are always checked. It can only help to swap the order.